// cactus-react-native/lib/commonjs/telemetry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Telemetry = void 0;
var _reactNative = require("react-native");
var _projectId = require("./projectId");
const packageJson = require('../package.json');
class Telemetry {
  static instance = null;
  constructor(config) {
    this.config = {
      table: 'telemetry',
      ...config
    };
  }
  static getFilename(path) {
    try {
      return path.split('/').pop() || path.split('\\').pop() || 'unknown';
    } catch {
      return 'unknown';
    }
  }
  static autoInit() {
    if (!Telemetry.instance) {
      Telemetry.instance = new Telemetry({
        supabaseUrl: 'https://vlqqczxwyaodtcdmdmlw.supabase.co',
        supabaseKey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InZscXFjenh3eWFvZHRjZG1kbWx3Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTE1MTg2MzIsImV4cCI6MjA2NzA5NDYzMn0.nBzqGuK9j6RZ6mOPWU2boAC_5H9XDs-fPpo5P3WZYbI' // Anon!
      });
    }
  }
  static init(config) {
    if (!Telemetry.instance) {
      Telemetry.instance = new Telemetry(config);
    }
  }
  static track(payload, options, deviceMetadata) {
    if (!Telemetry.instance) {
      Telemetry.autoInit();
    }
    Telemetry.instance.trackInternal(payload, options, deviceMetadata);
  }
  static error(error, options) {
    if (!Telemetry.instance) {
      Telemetry.autoInit();
    }
    Telemetry.instance.errorInternal(error, options);
  }
  trackInternal(payload, options, deviceMetadata) {
    const record = {
      project_id: _projectId.PROJECT_ID,
      device_id: deviceMetadata?.deviceId,
      device_manufacturer: deviceMetadata?.make,
      device_model: deviceMetadata?.model,
      os: _reactNative.Platform.OS === 'ios' ? 'iOS' : 'Android',
      os_version: _reactNative.Platform.Version.toString(),
      framework: 'react-native',
      framework_version: packageJson.version,
      telemetry_payload: payload,
      timestamp: new Date().toISOString(),
      model_filename: Telemetry.getFilename(options.model || ''),
      n_ctx: options.n_ctx,
      n_gpu_layers: options.n_gpu_layers
    };
    this.sendRecord(record).catch(() => {});
  }
  errorInternal(error, options) {
    const errorPayload = {
      message: error.message,
      stack: error.stack,
      name: error.name
    };
    const record = {
      project_id: _projectId.PROJECT_ID,
      os: _reactNative.Platform.OS === 'ios' ? 'iOS' : 'Android',
      os_version: _reactNative.Platform.Version.toString(),
      framework: 'react-native',
      framework_version: packageJson.version,
      error_payload: errorPayload,
      timestamp: new Date().toISOString(),
      model_filename: Telemetry.getFilename(options.model || ''),
      n_ctx: options.n_ctx,
      n_gpu_layers: options.n_gpu_layers
    };
    this.sendRecord(record).catch(() => {});
  }
  async sendRecord(record) {
    await globalThis.fetch(`${this.config.supabaseUrl}/rest/v1/${this.config.table}`, {
      method: 'POST',
      headers: {
        'apikey': this.config.supabaseKey,
        'Authorization': `Bearer ${this.config.supabaseKey}`,
        'Content-Type': 'application/json',
        'Prefer': 'return=minimal'
      },
      body: JSON.stringify([record])
    });
  }
}
exports.Telemetry = Telemetry;
