// cactus-react-native/lib/commonjs/tools.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tools = void 0;
exports.parseAndExecuteTool = parseAndExecuteTool;
class Tools {
  tools = new Map();
  add(func, description, parameters) {
    this.tools.set(func.name, {
      func,
      description,
      parameters,
      required: Object.entries(parameters).filter(([_, param]) => param.required).map(([key, _]) => key)
    });
    return func;
  }
  getSchemas() {
    return Array.from(this.tools.entries()).map(([name, {
      description,
      parameters,
      required
    }]) => ({
      type: "function",
      function: {
        name,
        description,
        parameters: {
          type: "object",
          properties: parameters,
          required
        }
      }
    }));
  }
  async execute(name, args) {
    const tool = this.tools.get(name);
    if (!tool) throw new Error(`Tool ${name} not found`);
    return await tool.func(...Object.values(args));
  }
}
exports.Tools = Tools;
async function parseAndExecuteTool(result, tools) {
  if (!result.tool_calls || result.tool_calls.length === 0) {
    return {
      toolCalled: false
    };
  }
  try {
    const toolCall = result.tool_calls[0];
    if (!toolCall) {
      return {
        toolCalled: false
      };
    }
    const toolName = toolCall.function.name;
    const toolInput = JSON.parse(toolCall.function.arguments);
    const toolOutput = await tools.execute(toolName, toolInput);
    return {
      toolCalled: true,
      toolName,
      toolInput,
      toolOutput
    };
  } catch (error) {
    return {
      toolCalled: false
    };
  }
}
